test/test_dir.rs

```diff
@@ -0,0 +1,46 @@
+extern crate nix;
+extern crate tempdir;
+
+use nix::dir::{Dir, Type};
+use nix::fcntl::OFlag;
+use nix::sys::stat::Mode;
+use std::fs::File;
+use self::tempdir::TempDir;
+
+#[test]
+fn read() {
+    let tmp = TempDir::new("nix-dir-read").unwrap();
+    File::create(&tmp.path().join("foo")).unwrap();
+    ::std::os::unix::fs::symlink("foo", tmp.path().join("bar")).unwrap();
+    let mut dir = Dir::open(tmp.path(), OFlag::O_DIRECTORY | OFlag::O_RDONLY | OFlag::O_CLOEXEC,
+                            Mode::empty()).unwrap();
+    let mut entries: Vec<_> = dir.iter().map(|e| e.unwrap()).collect();
+    entries.sort_by(|a, b| a.file_name().cmp(b.file_name()));
+    let entry_names: Vec<_> = entries
+        .iter()
+        .map(|e| e.file_name().to_str().unwrap().to_owned())
+        .collect();
+    assert_eq!(&entry_names[..], &[".", "..", "bar", "foo"]);
+
+    // Check file types. The system is allowed to return DT_UNKNOWN (aka None here) but if it does
```

Comment:
But why would it return `DT_UNKNOWN`? Does it do this in practice on any systems?

Reply:
Yes, it does. From the Linux `readdir(3)` manpage:

```
              Currently,  only  some  filesystems  (among them: Btrfs, ext2, ext3, and ext4) have
              full support for returning the file type in d_type.  All applications must properly
              handle a return of DT_UNKNOWN.
```